services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_volume_removal.py

```diff
@@ -0,0 +1,158 @@
+
+log = logging.getLogger(__name__)
+
+SERVICE_FINISHED_STATES: set[str] = {
+    "complete",
+    "failed",
+    "shutdown",
+    "rejected",
+    "orphaned",
+    "remove",
+}
+
+
+async def remove_volumes_from_node(
+    dynamic_sidecar_settings: DynamicSidecarSettings,
+    volume_names: list[str],
+    docker_node_id: str,
+    *,
+    volume_removal_attempts: int = 15,
+    sleep_between_attempts_s: int = 2,
+) -> bool:
+    """
+    Starts a service at target docker node which will remove
+    all entries in the `volumes_names` list.
+    """
+
+    async with docker_client() as client:
+        # When running docker-dind makes sure to use the same image as the
+        # underlying docker-engine.
+        # Will start a container with the same version across the entire cluster.
+        # It is safe to assume the local docker engine version will be
+        # the same as the one on the targeted node.
+        version_request = await client._query_json(  # pylint: disable=protected-access
+            "version", versioned_api=False
+        )
+        docker_version: DockerVersion = parse_obj_as(
+            DockerVersion, version_request["Version"]
+        )
+
+        # Timeout for the runtime of the service is calculated based on the amount
+        # of attempts required to remove each individual volume,
+        # in the worst case scenario when all volumes are do not exit.
+        volume_removal_timeout_s = volume_removal_attempts * sleep_between_attempts_s
+        service_timeout_s = volume_removal_timeout_s * len(volume_names)
+
+        service_spec = spec_volume_removal_service(
+            dynamic_sidecar_settings=dynamic_sidecar_settings,
+            docker_node_id=docker_node_id,
+            volume_names=volume_names,
+            docker_version=docker_version,
+            volume_removal_attempts=volume_removal_attempts,
+            sleep_between_attempts_s=sleep_between_attempts_s,
+            service_timeout_s=service_timeout_s,
+        )
+
+        volume_removal_service = await client.services.create(
+            **jsonable_encoder(service_spec, by_alias=True, exclude_unset=True)
+        )
+
+        service_id = volume_removal_service["ID"]
+        try:
+            async for attempt in AsyncRetrying(
+                stop=stop_after_delay(service_timeout_s),
+                wait=wait_fixed(0.5),
+                retry=retry_if_exception_type(_RetryError),
+                reraise=True,
+            ):
+                with attempt:
+                    tasks = await client.tasks.list(filters={"service": service_id})
+                    # NOTE: the service will have at most 1 task, since there is no restart
+                    # policy present
+                    if len(tasks) != 1:
+                        # Sometimes swarm did not mange to create a task after the service
+                        # is created. In that case a retry is triggered
+                        raise _RetryError(
```

Comment:
This comes from tenacity documentation (what about using what they already have? _RetryError is not necessary I think if the idea is to force trigger a retry.):

t’s also possible to retry explicitly at any time by raising the TryAgain exception:
```
@retry
def do_something():
    result = something_else()
    if result == 23:
       raise TryAgain

```